File: test-app/OrangeCandy.TestApp/Result.cs
namespace OrangeCandy.TestApp;

public sealed record Result<T>
{
    public T? Value { get; }
    public string? Error { get; }
    public bool IsSuccess => Error is null;

    private Result(T value) => Value = value;
    private Result(string error) => Error = error;

    public static Result<T> Success(T value) => new(value);
    public static Result<T> Failure(string error) => new(error);

    public Result<TNext> Map<TNext>(Func<T, TNext> transform) =>
        IsSuccess ? Result<TNext>.Success(transform(Value!)) : Result<TNext>.Failure(Error!);

    public async Task<Result<TNext>> Bind<TNext>(Func<T, Task<Result<TNext>>> transform) =>
        IsSuccess ? await transform(Value!) : Result<TNext>.Failure(Error!);

    public T Unwrap() => IsSuccess ? Value! : throw new InvalidOperationException(Error);
}

File: mcp-server/result.go
package main

type Result[T any] struct {
	value T
	err   error
	ok    bool
}

func Ok[T any](value T) Result[T] {
	return Result[T]{value: value, ok: true}
}

func Fail[T any](err error) Result[T] {
	return Result[T]{err: err, ok: false}
}

func (r Result[T]) IsOk() bool    { return r.ok }
func (r Result[T]) Value() T      { return r.value }
func (r Result[T]) Error() error  { return r.err }

func (r Result[T]) Unwrap() (T, error) {
	if r.ok {
		return r.value, nil
	}
	return r.value, r.err
}

func Then[T any, U any](r Result[T], f func(T) Result[U]) Result[U] {
	if !r.ok {
		return Fail[U](r.err)
	}
	return f(r.value)
}
